Improved error messaging when Process.register/2 fails

efexen opened the pull request:
Hey :wave: 

I've done a bit of work to improve the error messages when Process.register/2 fails.

Currently any of the failure cases will result in the following:

 ```
** (ArgumentError) argument error
             :erlang.register(:test, #PID<0.170.0>)
    (elixir) lib/process.ex:414: Process.register/2
```

Reading through the documentation it appears there are a some cases where we could make the messages more helpful.

First is the case where we are attempting to register a process which has already finished. We can use `Process.alive?/1` to determine whether the process is still alive and provide an enhanced message:

```
iex(1)> pid = spawn(fn() -> 1 end)
#PID<0.85.0>
iex(2)> Process.register(pid, :test)
** (ArgumentError) Process for #PID<0.85.0> is no longer alive
             :erlang.register(:test, #PID<0.85.0>)
    (elixir) lib/process.ex:424: Process.register/2
```

Second is the case where the name has already been registered for another process. We can use `Process.whereis/1` to determine whether this is the case and provide an enhanced message:

```
iex(1)> Process.register(self(), :test)
true
iex(2)> Process.register(self(), :test)
** (ArgumentError) Process with pid #PID<0.81.0> is already registered for name :test
             :erlang.register(:test, #PID<0.81.0>)
    (elixir) lib/process.ex:424: Process.register/2
```

Third is the case where a process is already registered with another name. We can find the existing registered name in the keyword list returned by `Process.info/1` and provide an enhanced message:

```
iex(1)> Process.register(self(), :test)
true
iex(2)> Process.register(self(), :tester)
** (ArgumentError) process #PID<0.81.0> is already registered with name :test
    (elixir) lib/process.ex:427: Process.register(#PID<0.81.0>, :tester)
```

I think all of these messages provide useful information that should make debugging the errors easier. I've also included a fallback option in case that the process is still alive and there isn't another process registered by that name it will use the original error message.

I've kept the errors as type `ArgumentError` so as not to break any code that attempts to rescue these errors, do you think this is preferable to introducing a new error type?

I'd love to hear your thoughts and whether we could work on this further to get it merged to provide better messages.

I wasn't entirely sure what to do around things such as testing for these errors so for now I've included them in the exception tests.

I have some further ideas for error message enhancements but didn't want to bundle them all here upfront without hearing some feedback first.

Thanks,
Ville

ericmj commented on lib/elixir/lib/exception.ex:
```diff
@@ -775,6 +775,25 @@ defmodule UnicodeConversionError do
+          " is already registered for name #{ inspect(name) }"
+      _ ->
+        original_message
+    end
```
It's not directly which `else` clause relates to which expression in the `with`. I wouldn't use `with` here.

efexen replied:
Hey Eric,

That was my concern with this, what would you recommend here instead? Had hard time coming up with what felt like Elixiry way of doing these 2 checks.

Thanks,
Ville

ericmj replied:
I think `cond` would work nicely:

```
cond do
  Process.alive? ->
  Process.whereis ->
  true ->
end
```

ericmj commented on lib/elixir/lib/exception.ex:
```diff
@@ -775,6 +775,25 @@ defmodule UnicodeConversionError do
+    end
+  end
+
+end
```
Remove the newlines before and after the function definition.

ericmj commented on lib/elixir/test/elixir/exception_test.exs:
```diff
@@ -335,6 +335,33 @@ defmodule ExceptionTest do
+      assert String.contains?(error.message, "is already registered")
+    end
+
+  end
```
Remove the spaces inside the `describe` here too.

ericmj commented:
There is a race condition between `:erlang.register` and the process alive/whereis checks. Not sure if it would cause any extra confusion if the race condition is hit though.

ericmj commented on lib/elixir/lib/process.ex:
```diff
@@ -420,7 +420,13 @@ defmodule Process do
+    rescue
+      error in ArgumentError ->
+        message = Process.RegisterError.message(pid, name, error.message)
+        reraise %{ error | message: message }, System.stacktrace
```
Remove spaces inside `%{ }` here too.

josevalim replied:
We are building an exception and not really using it. We also don't care about the stacktrace, so I would rewrite this as:

```elixir
  def register(pid, name) when not name in [nil, false, true] and is_atom(name) and is_pid(pid) do
    :erlang.register(pid, name)
  catch
    :error, :badarg ->
      message = "..."
      :erlang.error ArgumentError.exception(message), [pid, name]
  end
```

Where `message = "..."` is the message that you have written. This also keep it semantically closer to what Erlang does, without leaking the :erlang.register call in the stacktrace.

ericmj commented on lib/elixir/test/elixir/exception_test.exs:
```diff
@@ -335,6 +335,33 @@ defmodule ExceptionTest do
+    test "registering duplicate name" do
+      error = assert_raise ArgumentError, fn ->
+        Process.register(self(), :my_process)
+        Process.register(self(), :my_process)
```
Do the first register and the spawn+exit outside the `assert_raise`.

ericmj commented on lib/elixir/test/elixir/exception_test.exs:
```diff
@@ -335,6 +335,33 @@ defmodule ExceptionTest do
+
+      if is_pid(Process.whereis(:my_process)) do
+        Process.unregister(:my_process)
+      end
```
There shouldn't be any need for this. The process will die after the test finishes.

josevalim commented on lib/elixir/lib/exception.ex:
```diff
@@ -775,6 +775,25 @@ defmodule UnicodeConversionError do
 
+defmodule Process.RegisterError do
+
+  def message(pid, name, original_message) do
```
Since we never raise this. I would move this function directly to the Process module and keep it private.

ericmj commented on lib/elixir/lib/process.ex:
```diff
@@ -548,4 +552,16 @@ defmodule Process do
+
+  defp process_register_error(pid, name) do
+    case {alive?(pid), whereis(name)} do
+      { false, _ } ->
```
Don't use spaces inside braces :)

efexen replied:
Thanks :smile: fixed again

efexen commented:
Thanks for the feedback and recommendations, I believe my latest commit addresses them.

@ericmj Thanks for the feedback on code style, I think I've addressed all of them now, I ended up going with a case instead as the cond would have had weird bits like negating the first bit and trying to capture things, hope that's ok

@josevalim Thanks for the recommendation, I moved it to a private function in the process module and moved the tests as well as they didn't seem to make sense in the exception file anymore?

My original reasoning behind reraise was to preserve existing the functionality which refers to the :erlang.register part although I guess it doesn't really provide any value

lexmag commented on lib/elixir/lib/process.ex:
```diff
@@ -548,4 +552,16 @@ defmodule Process do
+      {false, _} ->
+        "Process for #{inspect(pid)} is no longer alive"
+      {true, existing_pid} when is_pid(existing_pid) ->
+        "Process with pid #{inspect(existing_pid)}" <>
```
In Logger we use `"process `#{inspect(existing_pid)}`" formatting.

efexen replied:
To make this more consistent maybe the error message could be swapped around to be "The name #{inspect(name)} is already registered to process #{inspect(existing_pid)}", what do you think?

lexmag commented on lib/elixir/test/elixir/process_test.exs:
```diff
@@ -132,6 +132,29 @@ defmodule ProcessTest do
+      Process.register(self(), :my_process)
+
+      error = assert_raise ArgumentError, fn ->
+        Process.register(self(), :my_process)
```
Hard to say what we actually is testing here since 2 errors happen: process can have only one registered name and name is already taken.

efexen replied:
Ooooh, I wasn't aware that a process can only have a single name registered, do you think it'd make sense to test for that as a separate case as well here?

efexen commented on lib/elixir/lib/process.ex:
```diff
@@ -548,4 +552,16 @@ defmodule Process do
+  defp process_register_error(pid, name) do
+    case {alive?(pid), whereis(name)} do
+      {false, _} ->
+        "Process for #{inspect(pid)} is no longer alive"
```
How about rewording this to be `"process #{inspect(pid)} is no longer alive"`, would that then be more consistent with the Logger formatting?

lexmag replied:
Looks 👍.

efexen commented on lib/elixir/test/elixir/process_test.exs:
```diff
@@ -132,6 +132,29 @@ defmodule ProcessTest do
+        Process.register(self(), :my_process)
+      end
+
+      assert String.contains?(error.message, "is already registered")
```
Good point I'll change both of them to do this instead :+1:

fishcakez commented:
If we catch an `ArgumentError` and then try to detect which reason from many it was we are open to race conditions as the registry and processes involved may have changed as it is not atomic with the `:erlang` call. We can only improve message by including all possibilities for each function. It might be possible to reduce this slightly by looking at the type involved.

fishcakez commented:
The reason provide by `:noproc` in `Exception` might be useful here.

josevalim commented:
@fishcakez if the process is alive though, we do have a better idea of what happened. If the process is dead, then all bets are off. We can amend those error messages later on after merged.

fishcakez commented:
If the process is alive we do not know that it was alive when the call was
made. It is possible for a pid to appear due to binary_to_term and pid/3 in
the shell.

If we could make that inference we don't know if it failed because the
process was already named or the name was registered to another process. So
it is not that helpful anyway.

On Saturday, 1 October 2016, José Valim <notifications@github.com> wrote:

> @fishcakez <https://github.com/fishcakez> if the process is alive though,
> we do have a better idea of what happened. If the process is dead, then all
> bets are off. We can amend those error messages later on after merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/5278#issuecomment-250917600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6JTeDSrVqxWKJzzwRlf9k_fDH96Gytks5qvnlkgaJpZM4KLgXA>
> .
>

lexmag commented:
I agree with @fishcakez, seems not much we can improve there.

efexen commented:
Hey @fishcakez you're absolutely right that there exists possibility of race conditions here due to the actions not being atomic but as far as I can tell hitting one of the race conditions would only result in the fallback error message which was the existing behaviour anyway?

Do you have anymore details about how I could check for `:noproc` to increase the likelihood of serving the enhanced error? It looks like the call to `:erlang.register/2` always fails with `:badarg` regardless of the reason.

efexen commented:
@lexmag I've changed the tests to use the clearer `assert_raise/3` function, thanks for the tip :+1:

I also managed to add a case to provide a different error message when the process we are trying to register has already been registered with another as per your recommendation. I've amended the original PR description and added a new test case to test for this as well, let me know what you think.

fishcakez commented:
@efexen sorry I wasn't very clear. I was suggesting we catch the `:badarg` and raise an `ArgumentError` with a constant message. The message being similar to the one in `Exception` when the exit reason is `:noproc`: https://github.com/elixir-lang/elixir/blob/97796f4d558826380904c32cadc30a335d99a50d/lib/elixir/lib/exception.ex#L242-L244.

I was saying that I don't think we should not use `Process.alive?` or `Process.whereis` because there is always a possibility to produce an incorrect error due to race conditions. Rather we should improve the message by listing all the possibilities and never falling back. It possible to can raise separate some errors by using `node/1`, `is_atom/1`, `is_pid/1`.

efexen commented:
@fishcakez I think I get it, sorry some of the double negatives are causing a bit of confusion for me.

What you're suggesting is obviously very radically different to what I've done here, should I close this PR and open a new one with a single all encompassing error message instead? Is there someone to consult in regards to the wording or language as it's likely to be quite a wordy one :smile:

fishcakez commented:
@efexen lets wait for @josevalim to comment

fishcakez commented:
@efexen sorry I edited my reply to remove the double negative that slipped in by mistake. I was looking for a double negative in exception.ex :)

josevalim commented:
Ok, let's go with a single message, proposal:

```elixir
"could not register the process #{inspect pid} with name #{inspect name}. Or the process is not alive, or the name is already taken, or the process has already been named"
```

efexen commented:
Thanks @josevalim 

I'll create a new PR with just that error message change

josevalim commented:
@efexen :heart:
